torchft/process_group_test.py

```diff
@@ -87,98 +110,173 @@ def check_tensors(arg: Any) -> None:  # pyre-ignore[2]
-        )
-
-        store_addr = f"localhost:{store.port}/prefix"
-        pg = ProcessGroupGloo()
-        pg.configure(store_addr, 0, 1)
-
-        self.assertEqual(pg.size(), 1)
-
-        _test_pg(pg)
-
-        m = nn.Linear(3, 4)
-        m = torch.nn.parallel.DistributedDataParallel(m, process_group=pg)
-        m(torch.rand(2, 3))
+@skipUnless(torch.cuda.is_available(), "needs CUDA")
+class NCCLTests(TestCase):
+    collectives = [
+        "allreduce",
+        "allgather",
+        "broadcast",
+        "broadcast_one",
+    ]
 
-    def test_gloo_timeout(self) -> None:
-        store = TCPStore(
+    def setUp(self) -> None:
+        self.store = TCPStore(
             host_name="localhost", port=0, is_master=True, wait_for_workers=False
         )
+        self.store_addr = f"localhost:{self.store.port}/prefix"
 
-        store_addr = f"localhost:{store.port}/prefix"
-        pg = ProcessGroupGloo(timeout=timedelta(seconds=0.01))
-        with self.assertRaisesRegex(
-            RuntimeError, "(timeout after 10ms|Socket Timeout)"
-        ):
-            pg.configure(store_addr, 0, 2)
-
-    # pyre-fixme[56]: Pyre was not able to infer the type of argument
-    @skipUnless(torch.cuda.is_available(), "needs CUDA")
-    def test_nccl(self) -> None:
-        store = TCPStore(
-            host_name="localhost", port=0, is_master=True, wait_for_workers=False
-        )
+    @parameterized.expand(collectives)
+    def test_nccl(self, collective: str) -> None:
         device = "cuda"
 
-        store_addr = f"localhost:{store.port}/prefix"
         pg = ProcessGroupNCCL()
-        pg.configure(store_addr, 0, 1)
+        pg.configure(self.store_addr, 0, 1)
 
         self.assertEqual(pg.size(), 1)
 
-        _test_pg(pg, torch.tensor([2], device=device))
+        run_collective(
+            pg=pg,
+            collective=collective,
+            example_tensor=torch.tensor([2], device=device),
+        )
 
         m = nn.Linear(3, 4).to(device)
         m = torch.nn.parallel.DistributedDataParallel(m, process_group=pg)
         m(torch.rand(2, 3, device=device))
 
         # reconfigure
-        store_addr = f"localhost:{store.port}/prefix2"
+        store_addr = f"localhost:{self.store.port}/prefix2"
         pg.configure(store_addr, 0, 1)
 
-        _test_pg(pg, torch.tensor([2], device=device))
+        run_collective(
+            pg=pg,
+            collective=collective,
+            example_tensor=torch.tensor([2], device=device),
+        )
 
         torch.cuda.synchronize()
 
-    def test_baby_gloo(self) -> None:
-        store = TCPStore(
+    @parameterized.expand(collectives)
+    def test_baby_nccl_apis(self, collective: str) -> None:
+        # set to 1 if more than >=2 gpus
+        device_id = 1 % torch.cuda.device_count()
+        torch.cuda.set_device(device_id)
+
+        pg = ProcessGroupBabyNCCL(timeout=timedelta(seconds=10))
+        try:
+            pg.configure(self.store_addr, 0, 1)
```

Comment:
This seems really slow -- how fast does this run? Launching the subprocess is pretty slow so would actually prefer to run these all on the same PG

If you want prettier printing we can use subtests?

i.e.

```
for collective in collectives:
    with self.subTest(collective=collective):
        ...
```

Reply:
Good callout, with parameterized it took ~36s, without it took ~16s. I've removed parameterized for now.